operators/pkg/controller/elasticsearch/elasticsearch_controller.go

```diff
@@ -283,3 +297,78 @@ func (r *ReconcileElasticsearch) finalizersFor(
 		http.DynamicWatchesFinalizer(r.dynamicWatches, es.Name, esname.ESNamer),
 	}
 }
+
+// reconcileCompatibility determines if this controller is compatible with a given resource by examining the controller version annotation
+// controller versions 0.9.0+ cannot reconcile resources created with earlier controllers, so this lets our controller skip those resources until they can be manually recreated
+// if an object does not have an annotation, it will determine if it is a new object or if it has been previously reconciled by an older controller version, as this annotation
+// was not applied by earlier controller versions. it will update the object's annotations indicating it is incompatible if so
+func (r *ReconcileElasticsearch) reconcileCompatibility(es *elasticsearchv1alpha1.Elasticsearch) (bool, error) {
+	annExists := es.Annotations != nil && es.Annotations[annotation.ControllerVersionAnnotation] != ""
+
+	// if the annotation does not exist, it might indicate it was reconciled by an older controller version that did not add the version annotation,
+	// in which case it is incompatible with the current controller, or it is a brand new resource that has not been reconciled by any controller yet
+	if !annExists {
+		exist, err := r.checkExistingResources(es)
+		if err != nil {
+			return false, err
+		}
+		if exist {
+			log.Info("Resource was previously reconciled by incompatible controller version and missing annotation, adding annotation", "controller_version", r.OperatorInfo.BuildInfo.Version, "namespace", es.Namespace, "es_name", es.Name)
+			err = annotation.UpdateControllerVersion(r.Client, es, "0.8.0-UNKNOWN")
+			return false, err
+		}
+		// no annotation exists and there are no existing resources, so this has not previously been reconciled
+		err = annotation.UpdateControllerVersion(r.Client, es, r.OperatorInfo.BuildInfo.Version)
+		return true, err
+	}
+
+	// if we have an annotation we need to check if it is compatible
+	currentVersion, err := semver.NewVersion(es.Annotations[annotation.ControllerVersionAnnotation])
+	if err != nil {
+		return false, errors.Wrap(err, "Error parsing current version on resource")
+	}
+	minVersion, err := semver.NewVersion("0.9.0-ALPHA")
```

Comment:
Not sure if it is important, but could we avoid hardcoding a version here by saying version compatibility is given if major and minor are the same?

Also we have 

https://github.com/elastic/cloud-on-k8s/blob/f7afa848751ff787e2524d2829d1b54307755341/operators/pkg/controller/common/version/version.go#L15-L20

Reply:
Ah I missed that version package. I am kind of okay with hardcoding package incompatibility since it should be the exception rather than the rule imo. Like in the future if we go from controller 1.2 -> 1.3 because we added a feature, we shouldn't make the assumption right now that controller 1.3 cannot reconcile existing CRs. Definitely possible I'm missing something though.